coursera/api.py

```diff
@@ -455,6 +464,116 @@ def extract_links_from_exam(self, exam_id):
                 logging.exception('Could not download exam %s: %s', exam_id, exception)
             return None
 
+    def _get_notebook_folder(self, url, jupyterId, **kwargs):
+
+        supplement_links = {}
+
+        url = url.format(**kwargs)        
+        reply = get_page(
+            self._session,
+            url,
+            json=True
+        )
+
+        headers = self._auth_headers_with_json()
+
+        for content in reply['content']:
+            
+            if content['type'] == 'directory':
+                a = self._get_notebook_folder(OPENCOURSE_NOTEBOOK_TREE, jupyterId, jupId=jupyterId, path=content['path'], timestamp=int(time.time()))
+                supplement_links.update(a)
+            
+            elif content['type'] == 'file':
+                tmpUrl = OPENCOURSE_NOTEBOOK_DOWNLOAD.format(path=content['path'], jupId=jupyterId, timestamp=int(time.time()))
+                filename, extension = os.path.splitext(clean_url(tmpUrl))
+                
+                head, tail = os.path.split(content['path'])
+                
+                if os.path.isdir(self._course_name + "/notebook/" + head + "/") == False:
+                    logging.info('Creating [{}] directories...'.format(head))
+                    os.makedirs(self._course_name + "/notebook/" + head + "/")
+                
+                r = requests.get(tmpUrl.replace(" ", "%20"), cookies=self._session.cookies)
+                if os.path.exists(self._course_name + "/notebook/" + head + "/" + tail) == False:
+                    logging.info('Downloading {} into {}'.format(tail, head))
+                    with open(self._course_name + "/notebook/" + head + "/" + tail, 'wb+') as f:
+                        f.write(r.content)
+                else:
+                    logging.info('Skipping {}... (file exists)'.format(tail))
+
+
+                if not str(extension[1:]) in supplement_links:
+                    supplement_links[str(extension[1:])] = []
+                
+                supplement_links[str(extension[1:])].append((tmpUrl.replace(" ", "%20"), filename))
```

Comment:
Consider using `urlencode` and save the result into a variable as you're using it at least twice in this block.